execution_engine/src/storage/trie/merkle_proof.rs

```diff
@@ -0,0 +1,323 @@
+    ///
+    /// The steps in this function reflect `operations::rehash`.
+    pub fn compute_state_hash(&self) -> Result<Blake2bHash, Error> {
+        let mut hash = {
+            let leaf_bytes = Trie::leaf(self.key, self.value).to_bytes()?;
+            Blake2bHash::new(&leaf_bytes)
+        };
+
+        for (proof_step_index, proof_step) in self.proof_steps.iter().enumerate() {
+            let pointer = if proof_step_index == 0 {
+                Pointer::LeafPointer(hash)
+            } else {
+                Pointer::NodePointer(hash)
+            };
+            let proof_step_bytes = match proof_step {
+                TrieMerkleProofStep::Node {
+                    hole_index,
+                    indexed_pointers_with_hole,
+                } => {
+                    let hole_index = *hole_index;
+                    if hole_index as usize > RADIX {
+                        let proof_step_index = proof_step_index.try_into().expect(USIZE_EXCEEDS_U8);
+                        return Err(Error::HoleIndexOutOfRange {
+                            hole_index,
+                            proof_step_index,
+                        });
+                    }
+                    let mut indexed_pointers = indexed_pointers_with_hole.to_owned();
+                    indexed_pointers.push((hole_index, pointer));
+                    Trie::<K, V>::node(&indexed_pointers).to_bytes()?
+                }
+                TrieMerkleProofStep::Extension { affix } => {
+                    Trie::<K, V>::extension(affix.to_owned(), pointer).to_bytes()?
+                }
+            };
+            hash = Blake2bHash::new(&proof_step_bytes);
+        }
+        Ok(hash)
+    }
+}
+
+impl<K, V> ToBytes for TrieMerkleProof<K, V>
+where
+    K: ToBytes,
+    V: ToBytes,
+{
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        let mut ret: Vec<u8> = bytesrepr::allocate_buffer(self)?;
+        ret.extend(self.key.to_bytes()?);
+        ret.extend(self.value.to_bytes()?);
+        ret.extend(self.proof_steps.to_bytes()?);
+        Ok(ret)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.key.serialized_length()
+            + self.value.serialized_length()
+            + self.proof_steps.serialized_length()
+    }
+}
+
+impl<K, V> FromBytes for TrieMerkleProof<K, V>
+where
+    K: FromBytes,
+    V: FromBytes,
+{
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (key, rem): (K, &[u8]) = FromBytes::from_bytes(bytes)?;
+        let (value, rem): (V, &[u8]) = FromBytes::from_bytes(rem)?;
+        let (proof_steps, rem): (VecDeque<TrieMerkleProofStep>, &[u8]) =
+            FromBytes::from_bytes(rem)?;
+        Ok((
+            TrieMerkleProof {
+                key,
+                value,
+                proof_steps,
+            },
+            rem,
+        ))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Error {
```

Comment:
Couldn't we use `thiserror` here as well to avoid `Display for Error` and `impl From<bytesrepr::Error>`?

Reply:
I was unsure about introducing an additional usage of `thiserror`. Have we settled this as our chosen error handling library? CC @Fraser999

Reply:
We ended up being able to remove this error entirely